forum/qa-plugin/adventofcode-widget/src/adventofcode-widget.php

```diff
@@ -0,0 +1,200 @@
+                ]
+            ],
+            'buttons' => [
+                [
+                    'label' => qa_lang_html('adventofcode_widget/admin_save'),
+                    'tags' => 'name="adventofcode_widget_save"'
+                ],
+            ]
+        ];
+    }
+
+    private function is_enabled()
+    {
+        return qa_opt('adventofcode_widget_enabled') == 1;
+    }
+
+    private function should_update_results()
+    {
+        $date = qa_opt('adventofcode_widget_update_date');
+        $now = (new DateTime())->format('Y-m-d H');
+
+        return $date !== $now;
+    }
+
+    private function update_results()
+    {
+        $year = qa_opt('adventofcode_widget_year');
+        $leaderboard = qa_opt('adventofcode_widget_leaderboard_id');
+        $session = qa_opt('adventofcode_widget_session_id');
+
+        $aocResponse = $this->getAoCResponse($year, $leaderboard, $session);
+        $userLinks = $this->getUserLinks($year, $leaderboard, $session);
+
+        $content = $this->parseAocResponse($aocResponse, $userLinks);
+        if ($content) {
+            qa_opt('adventofcode_widget_content', $content);
+        }
+
+        qa_opt('adventofcode_widget_update_date', (new DateTime())->format('Y-m-d H'));
+
+        return true;
+    }
+
+    private function getAoCResponse($year, $leaderboard, $session)
+    {
+        $curl = curl_init();
+        curl_setopt($curl, CURLOPT_URL, "https://adventofcode.com/{$year}/leaderboard/private/view/{$leaderboard}.json");
+        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
+        curl_setopt($curl, CURLOPT_COOKIE, 'session=' . $session);
+        $aocResponse = curl_exec($curl);
+        $httpCode = curl_getinfo($curl, CURLINFO_HTTP_CODE);
+        curl_close($curl);
+
+        if ($httpCode !== 200) {
+            return null;
+        }
+
+        return $aocResponse;
+    }
+
+    private function getUserLinks($year, $leaderboard, $session)
+    {
+        $linksCurl = curl_init();
+        curl_setopt($linksCurl, CURLOPT_URL, "https://adventofcode.com/{$year}/leaderboard/private/view/{$leaderboard}");
+        curl_setopt($linksCurl, CURLOPT_RETURNTRANSFER, true);
+        curl_setopt($linksCurl, CURLOPT_COOKIE, 'session=' . $session);
+        $linksResponse = curl_exec($linksCurl);
+        $httpCode = curl_getinfo($linksCurl, CURLINFO_HTTP_CODE);
+        curl_close($linksCurl);
+
+        if ($httpCode !== 200) {
+            return null;
+        }
+
+        libxml_use_internal_errors(true);
+        $dom = new DOMDocument;
+        $userLinks = [];
+
+        if($dom->loadHTML($linksResponse, LIBXML_NOWARNING)) {
+            foreach($dom->getElementsByTagName('a') as $anchor) {
+                $potentialUserLink = $anchor->getAttribute('href');
+                $potentialUserName = $anchor->textContent;
+                $userLinks[$potentialUserName] = $potentialUserLink;
+            }
+        }
+
```

Comment:
Bazując na 62882ced1f388d6946e36bc162ef82ba777b2cd0 zalecałbym 

```suggestion
 libxml_use_internal_errors(false); 
```
https://www.php.net/manual/en/function.libxml-use-internal-errors.php#112123
https://www.php.net/manual/en/function.libxml-use-internal-errors.php#78236

Reply:
Jeśli nie zalecasz wyciszenia błędów to co proponujesz zrobić wtedy z warningami, które są rzucane?

Reply:
Zauważ jakie zmiany proponuję i w którym miejscu 🙈

Reply:
Ok, sorry, nie zwróciłem uwagi, że chodzi po prostu o przywrócenie poprzedniego stanu po. Z tym to nie ma problemu, dodałem